ic-os/components/misc/guestos-recovery/guestos-recovery-upgrader/guestos-recovery-upgrader.sh

```diff
@@ -0,0 +1,235 @@
+#!/bin/bash
+
+set -e
+
+# Perform a manual GuestOS upgrade from the HostOS
+
+# Constants for partitions and paths
+GRUB_PARTITION_NUM=2
+BOOT_PARTITION_A=4
+ROOT_PARTITION_A=5
+VAR_PARTITION_A=6
+BOOT_PARTITION_B=7
+ROOT_PARTITION_B=8
+VAR_PARTITION_B=9
+
+GUESTOS_DEVICE="/dev/hostlvm/guestos"
+IC_DOWNLOAD_BASE="https://download.dfinity.systems/ic"
```

Comment:
You should consider supporting our backup CDN here such that when AWS S3 fails we switch to downloading from CloudFront at `https://download.dfinity.network`.

Reply:
Great that will work. You could also consider just specifying a list of URLs like we do in proposals ([example](https://dashboard.internetcomputer.org/proposal/136740)). Then you could try them in sequence and stop at the first succeeding one. The advantage is that it's more consistent with how we handle upgrades and when we add more URLs you can simply add them to the list of URLs.